### ISSUE TITLE
[ALGO] Warp-level cooperative heterogeneous reduce-scatter with adaptive bucketing

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#48]
## Problem

Current reduce-scatter implementations in distributed training assume homogeneous data types and fixed bucket sizes, leading to suboptimal memory bandwidth utilization and excessive kernel launch overhead when dealing with mixed-precision gradients (FP32/BF16/FP8). Warp-level primitives like `__shfl_xor_sync` are underutilized for intra-block gradient aggregation.

## Approach

Implement a warp-level cooperative reduce-scatter that:
- Uses `cub::WarpReduce` as the inner reduction primitive with templated `ReductionOp`
- Introduces **adaptive bucketing**: at runtime, probe gradient norm distribution and dynamically merge small tensors into fused buckets (threshold: configurable, default 1 MB)
- Heterogeneous type dispatch via `cuda::std::variant<__half, __nv_bfloat16, float>` with compile-time specialization
- Exploit `nvcuda::wmma` for mixed-precision accumulation on SM8.0+

```cpp
template <typename InputIt, typename OutputIt, typename ReductionOp>
cudaError_t warp_reduce_scatter(
    InputIt  d_in,
    OutputIt d_out,
    size_t   num_elements,
    int      num_ranks,
    ReductionOp op,
    AdaptiveBucketConfig cfg,
    cudaStream_t stream = 0);
```

## API

```cpp
// Minimal usage
neuron::warp_reduce_scatter(
    grads.data_ptr(), output.data_ptr(),
    numel, world_size,
    neuron::Sum{},
    neuron::AdaptiveBucketConfig{.min_bucket_mb = 1, .max_buckets = 64},
    stream);
```

## Benchmarks

Target platforms: A100 (SM8.0), H100 (SM9.0), RTX 4090 (SM8.9)

| Config | Baseline (NCCL) | Target | Metric |
|---|---|---|---|
| 1B params, BF16, 8-GPU | 12.3 ms | < 9 ms | reduce-scatter latency |
| Bucket merge overhead | N/A | < 0.5 ms | adaptive bucketing cost |

Benchmark via `nvbench` with sweep over `[num_elements, dtype, world_size]`.

## References
- CCCL `cub::DeviceReduce` — [link](https://github.com/NVIDIA/cccl/blob/main/cub/cub/device/device_reduce.cuh)
- NCCL reduce-scatter primitive
- Megatron-LM gradient bucketing logic
